### PR TITLE
gc_worker: collect locks after applying

### PR DESCRIPTION
**Pull request by youjiali1995**
Signed-off-by: youjiali1995 <zlwgx1023@gmail.com>

### What problem does this PR solve?

Issue Number: close https://github.com/tikv/tikv/issues/8211
Problem Summary:
See the issue.

### What is changed and how it works?

What's Changed:

Notifying lock observer after applying logs.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

* Fix the issue that green GC may miss locks during transferring regions.

### CODE REVIEW

**Comment by sticnarf** [comment]
Sorry I haven't read the code carefully. But I wonder what if a `check_lock_observer` happens after `apply` and before `post_apply_query`. Will we get an incorrect check result and leave some locks unresolved?

**Comment by youjiali1995** [comment]
> Sorry I haven't read the code carefully. But I wonder what if a `check_lock_observer` happens after `apply` and before `post_apply_query`. Will we get an incorrect check result and leave some locks unresolved?

There is a document for it: https://docs.google.com/document/d/14eHWwtx3KTRpW8bzx876pp0V_1AAh19QRUJv8H2lM9Q/edit?disco=AAAAGs3z8qU

**Comment by sticnarf** [comment]
> > Sorry I haven't read the code carefully. But I wonder what if a `check_lock_observer` happens after `apply` and before `post_apply_query`. Will we get an incorrect check result and leave some locks unresolved?
> 
> There is a document for it: https://docs.google.com/document/d/14eHWwtx3KTRpW8bzx876pp0V_1AAh19QRUJv8H2lM9Q/edit?disco=AAAAGs3z8qU

Thanks, I see. 

By the way, could you make this document public?

File: tests/integrations/server/gc_worker.rs [hunk 1]
```diff
@@ -71,6 +71,7 @@ fn test_physical_scan_lock() {
 fn test_applied_lock_collector() {
     let mut cluster = new_server_cluster(0, 3);
     cluster.run();
+    cluster.pd_client.disable_default_operator();
```
Comment (sticnarf) [inline tests/integrations/server/gc_worker.rs:74]: What is this for?

Comment (youjiali1995) [reply tests/integrations/server/gc_worker.rs:74]: The test fails frequently when adds a new peer. The reason is PD's default operator already adds a new peer to the newly added store.

Comment (sticnarf) [reply tests/integrations/server/gc_worker.rs:74]: Got it. Thanks!

**Comment by youjiali1995** [comment]
@MyonKeminta @BusyJay PTAL, and it can close https://github.com/tikv/tikv/issues/8261.

File: components/raftstore/src/store/snap/io.rs [hunk 2]
```diff
@@ -156,12 +156,16 @@ where
...
-        batch.drain(..).try_for_each(|(k, v)| wb.put_cf(cf, &k, &v))
-    };
-
     let mut wb = db.write_batch();
+    let mut write_to_db =
+        |db: &E, batch: &mut Vec<(Vec<u8>, Vec<u8>)>| -> Result<(), EngineError> {
+            batch.iter().try_for_each(|(k, v)| wb.put_cf(cf, &k, &v))?;
+            db.write(&wb)?;
+            wb.clear();
+            callback(batch.drain(..).as_slice());
```
Comment (BusyJay) [inline components/raftstore/src/store/snap/io.rs:165]: Why not just `&batch`?

**Comment by BusyJay** [comment]
Can it be sent as a separate PR? They don't seem to be related.

File: components/raftstore/src/store/snap/io.rs (hunk 2 above)
Comment (youjiali1995) [reply components/raftstore/src/store/snap/io.rs:165]: Do you mean?
```suggestion
            callback(batch);
            batch.clear();
```

**Comment by youjiali1995** [comment]
> I'm worry it may lose updates supposing a thread is stuck after writing to rocksdb while before sending any updates to observer.

It won't, https://docs.google.com/document/d/14eHWwtx3KTRpW8bzx876pp0V_1AAh19QRUJv8H2lM9Q/edit?disco=AAAAGs3z8qU. @BusyJay

File: components/raftstore/src/store/snap/io.rs (hunk 2 above)
Comment (BusyJay) [reply components/raftstore/src/store/snap/io.rs:165]: Yes.

**Comment by youjiali1995** [comment]
@BusyJay @MyonKeminta PTAL!

File: src/server/gc_worker/applied_lock_collector.rs [hunk 3]
```diff
@@ -161,19 +163,20 @@ impl LockObserver {
...
+    fn post_apply_query(&self, _: &mut ObserverContext<'_>, cmd: &mut Cmd) {
+        fail_point!("notify_lock_observer_query");
+        let max_ts = self.state.load_max_ts();
+        if max_ts.is_zero() {
             return;
         }
 
-        let max_ts = self.state.load_max_ts();
-        if max_ts.is_zero() {
+        if !self.state.is_clean() {
```
Comment (MyonKeminta) [inline src/server/gc_worker/applied_lock_collector.rs:173]: Does it make any difference by flipping the order of max_ts check and state clean check?

Comment (youjiali1995) [reply src/server/gc_worker/applied_lock_collector.rs:173]: Normally, `state` is always clean. Changing the order can reduce one atomic load when green GC is not running.

**Pull request merged**